File: Binary-Search-Tree/BST-C#/src/Tree.cs
using System;

namespace src
{
  class Tree
  {
    public Tree(){this.root=null;}

    public void insert(string data,id)
    {
      this.root=insert(this.root,data,id);
    }

    public void delete(int id)
    {
      this.root=delete(this.root,id);
    }

    public void traverse()
    {
      traverse(this.root);
    }

    private Node insert(Node root,string data,int id)
    {
      if(root==null){return new Node(new Data(data,id));}

      if(id>root.data.id)
      {
        return root.right=insert(root.right,data,id);
      }
      else
      {
        return root.left=insert(root.left,data,id);
      }
    }

    private Node delete(Node root,int id)
    {
      if(root==null){return null;}

      if(id>root.data.id)
      {
        return root.right=delete(root.right,id);
      }
      else if(id<root.data.id)
      {
        return root.left=delete(root.left,id);
      }
      else
      {
        if(root.right==null||root.left==null)
        {
          return null;
        }
        else if(root.right==null)
        {
          return root.left;
        }
        else if(root.left==null)
        {
          return root.right;
        }
        else
        {
          Node temp=findMin(root.right);
          root.data.data=temp.data.data;
          root.data.id=temp.data.id;
          return root.right=delete(root.right,temp.data.id);
        }
      }
    }

    private void traverse(Node root)
    {
      if(root==null){return;}

      traverse(root.left);
      Console.WriteLine(root.data.data);
      traverse(root.right);
    }

    private Node findMin(Node root) 
    {
      if(root==null){return null;}

      while(root!=null&&root.left!=null)
      {
        root=root.left;
      }
      return root;
    }
  }
}

File: Binary-Search-Tree/BST-py/src/Tree.py
from Data.Node import Node
from Data.Data import Data

class Tree:
    def __init__(self):
        self.root:Node=None

    def insert(self,data:str,id:int)->None:
        self.root=self.__insert(self.root,data,id)

    def delete(self,id:int)->None:
        self.root=self.__delete(self.root,id)

    def view(self)->None:
        self.__view(self.root)

    def __insert(self,root:Node,data:str,id:int)->Node:
        if root is None:
            return Node(Data(data,id))

        if id>root.data.id:
            root.right=self.__insert(root.right,data,id)

        elif id<root.data.id:
            root.left=self.__insert(root.left,data,id)

        else:
            return None

    def __delete(self,root:Node,id:int)->Node:
        if root is None:
            return  None

        if id>root.data.id:
            root.right=self.__delete(root.right,id)
        elif id<root.data.id:
            root.left=self.__delete(root.left,id)
        else:
            if root.right is None:
                root.left
            elif root.left is None:
                root.right
            else:
                temp:Node=self.__findMin(root.right)
                root.data=temp.data
                root.right=self.__delete(root.right,temp.data.id)

    def __view(self,root:Node)->None:
        if root is None:
            return

        self.__view(root.left)
        print(root.data.data, end="")
        self.__view(root.right)


    def __findMin(self,root:Node)->Node:
        if root is None:
            return None

        while root is not None and root.left is not None:
            root=root.left
        return root
